contracts/FeeDisbursal/FeeDisbursal.sol

```diff
@@ -0,0 +1,269 @@
+        require(amount > 0, "NF_DISB_ZERO_AMOUNT");
+
+        // cast and check pro rata token
+        BasicSnapshotToken proRataToken = BasicSnapshotToken(decodeAddress(data));
+        uint256 snapshotId = proRataToken.currentSnapshotId() - 1;
+        require(proRataToken.totalSupplyAt(snapshotId) > 0, "");
+
+        // create a new disbursal entry
+        _disbursals[msg.sender].push(Disbursal({
+            recycleAfterTimestamp: block.timestamp,
+            amount: amount,
+            proRataToken: proRataToken,
+            snapshotId: snapshotId,
+            disburser: wallet
+        }));
+        //@TODO: add log message
+    }
+
+    // implementation of reclaimbale
+    function reclaim(IBasicToken token)
+        public
+        only(ROLE_RECLAIMER)
+    {   
+        // forbid reclaiming any or our platform payment tokens
+        // @TODO: move access control to Controller
+        require(!isDisbursableToken(address(token)));
+        Reclaimable.reclaim(token);
+    }
+
+    /// @notice helper to determine if the token at the given address is supported for disbursing
+    /// @param token address of token in question
+    function isDisbursableToken(address token)
+        public
+        view
+        returns (bool)
+    {   
+        // @TODO: migrate this to new, more flexible token registering in universe
+        if (token == address(UNIVERSE.etherToken())) return true;
+        if (token == address(UNIVERSE.euroToken())) return true;
+        return false;
+    }
+
+    ////////////////////////
+    // Private functions
+    ////////////////////////
+
+    /// @notice claim a token for an spender, returns the amount of tokens claimed
+    /// @param token address of the claimable token
+    /// @param spender address of the spender that will receive the funds
+    /// @param until until what index to claim to
+    function claimPrivate(address token, address spender, uint256 until)
+    public
+    returns (uint256 claimedAmount, uint256 lastIndex)
+    {
+        require(isDisbursableToken(token), "NF_DISB_UKNOWN_TOKEN");
+        (claimedAmount, lastIndex) = claimablePrivate(token, spender, until);
+
+        // mark spender disbursal progress
+        _disbursalProgress[token][spender] = lastIndex;
+
+        // do the actual token transfer
+        IERC223Token ierc223Token = IERC223Token(token);
+        ierc223Token.transfer(spender, claimedAmount, "");
+        //@TODO: add log message
+    }
+
+    /// @notice get the amount of tokens that can be claimed by a given spender
+    /// @param token address of the claimable token
+    /// @param spender address of the spender that will receive the funds
+    /// @param until until what index to claim to, use UINT256_MAX for all
+    function claimablePrivate(address token, address spender, uint256 until)
+    public
+    view
+    returns (uint256 claimableAmount, uint256 lastIndex)
+    {
+        require(isDisbursableToken(token), "NF_DISB_UKNOWN_TOKEN");
```

Comment:
here we should prevent reclaiming if `currentSnapshotId` is identical to snapshot in Disbursal struct. because for current snapshot id, balances and total supply are not sealed